### DISCUSSION TITLE
Concat Columns V 2.8

### DISCUSSION THREAD

**Discussion by julianR92** [rappasoft/laravel-livewire-tables#907] [Q&A] [answered]
hello colleagues how can I concatenate columns in this version I have something like this

```
 public function columns(): array
    {
        return [
            Column::make("Id", "id")
                ->sortable(),
                Column::make('Name')
                ->searchable()
               ->format(function($value, $column, $row) {
                    return $row->first_name . ' ' . $row->last_name;
                })
            
            
        ];
    }
```
and throws this error every time, what am I doing wrong?

`SQLSTATE[42S22]: Column not found: 1054 Unknown column 'users.name' in 'field list'
`

**Comment by lrljoe** [comment]
It'll look for the name of your column as the db field unless you specify a second parameter in the column make.
To demo this, use id as the second parameter for your name column, and your example will work.

Alternatively, use the label approach from the documentation

Column::make('My one off column') ->label( fn($row, Column $column) => $this->getSomeOtherValue($row, $column) ),

**Comment by julianR92** [comment]
hi friend I'm using the label documentation and it doesn't recognize the methods, I think my library has a problem

```
 public function columns(): array
    { 
        return [
            Column::make("Id", "id")
                ->sortable(),
                Column::make('Nombre','first_name')
                ->sortable(),
                Column::make('Apellidos','last_name')
                ->sortable(),
                Column::make('Usuario','email')
                ->sortable(),
                Column::make('My one off column') ->label( fn($row, Column $column) => $this->getSomeOtherValue($row, $column) ),

                Column::make('status','estado')                
                ->deselected(),// Hide this header on mobile
                Column::make('Acciones')
                ->label(
                    fn($row)=> view('livewire.users.users-actions',compact('row'))
                ),
                Column::make('Estado')
                ->label(
                    fn($row)=> view('livewire.users.estado-actions',compact('row'))
                ),              
                  
        ];
    }
```
and return

![image](https://user-images.githubusercontent.com/52473935/189491069-6e416c75-2f5f-493f-9a63-9f96136d5189.png)

my composer.json:

```
{
    "name": "laravel/laravel",
    "type": "project",
    "description": "The Laravel Framework.",
    "keywords": ["framework", "laravel"],
    "license": "MIT",
    "require": {
        "php": "^7.3|^8.0",
        "fruitcake/laravel-cors": "^2.0",
        "guzzlehttp/guzzle": "^7.4",
        "laravel/framework": "^9.0",
        "laravel/tinker": "^2.6",
        "laravelcollective/html": "^6.3",
        "livewire/livewire": "^2.8",
        "rappasoft/laravel-livewire-tables": "^2.8",
        "spatie/laravel-permission": "^5.5"
    },
    "require-dev": {
        "spatie/laravel-ignition": "^1.0",
        "fakerphp/faker": "^1.17",
        "laravel/sail": "^1.12",
        "mockery/mockery": "^1.4.4",
        "nunomaduro/collision": "^6.1",
        "phpunit/phpunit": "^9.5.10"
    },
```
thank you for your help

**Comment by lrljoe** [comment, marked as answer]
When you're using the following example:
Column::make('My one off column')
    ->label(
        fn($row, Column $column) => $this->getSomeOtherValue($row, $column)
    ),
If you're using $this->getSomeOtherValue, then that needs to be a function that will return the data that you want to display.  It can either be a function on the DataTable component (or a parent component).

If the function is in your model (for example), then you'd want to use $row->getSomeOtherValue() instead, passing the relevant parameters.

You can change getSomeOtherValue to whatever you choose to name your function.

As an example, in my TaskList table I have the following:

Column::make('Open Tasks')
                ->label(fn ($row) => $row->openTasks()),

This column calls a function that is named openTasks.  This function is stored within my TaskList model, which returns the number of open tasks that a particular tasklist has.

If you're still stuck, then please drop both the DataTable component, and the related Model somewhere that I can see them :)

**Reply by julianR92** [reply]
I understand what you are saying, but how can I concatenate my columns from the model?

**TABLE CLASS**

```
class UserTable extends DataTableComponent

{
    protected $model = User::class;
    public string $tableName = 'users';

    public function configure(): void
    {
        $this->setPrimaryKey('id');
        $this->setDefaultSort('id', 'asc');
        $this->setRefreshTime(5000);
    }

    public function columns(): array
    { 
        return [
            Column::make("Id", "id")
                ->sortable(),
                Column::make('Nombre','first_name')
                ->sortable(),
                Column::make('nombres')
                ->label(fn ($row) => $row->concatName()),
                Column::make('Usuario','email')
                ->sortable(),
                Column::make('status','estado')                
                ->deselected(),// Hide this header on mobile
                Column::make('Acciones')
                ->label(
                    fn($row)=> view('livewire.users.users-actions',compact('row'))
                ),
                Column::make('Estado')
                ->label(
                    fn($row)=> view('livewire.users.estado-actions',compact('row'))
                ),


                
               
                
                            
            
        ];
    }

    public function builder(): Builder
    {   
        
        $usuario = User::findOrFail(auth()->user()->id);
        if($usuario->getRoleNames()[0] == 'ADMIN'){
            return User::query()->where('role_id','!=' ,1)->where('role_id','!=',2)->leftjoin('roles', 'roles.id', 'users.role_id');   
         } else {
            return User::query()->leftjoin('roles', 'roles.id', 'users.role_id');
         }
    }

    
}
```
**MODEL**

```
class User extends Authenticatable
{
    use HasFactory, Notifiable, HasRoles, SoftDeletes;

    /**
     * The attributes that are mass assignable.
     *
     * @var array
     */
 /*    protected $fillable = [
        'name',
        'email',
        'password',
    ]; */
    protected $guarded=[];

    /**
     * The attributes that should be hidden for arrays.
     *
     * @var array
     */
    protected $hidden = [
        'password',
        'remember_token',
    ];

    /**
     * The attributes that should be cast to native types.
     *
     * @var array
     */
    protected $casts = [
        'email_verified_at' => 'datetime',
    ];

    public function concatName(){

        
    }

```
THANK YOU FOR YOUR HELP

**Reply by lrljoe** [reply]
To give you an example that will concatenate the name and email fields in your user model:

public function concatName(){
     return $this->name . " " . $this->email;
}

Or for first_name and last_name
public function concatName(){
     return $this->first_name . " " . $this->last_name;
}

As another quick tip for you, I do recommend that you consider localisation early on if you plan to scale your app at all, for example replacing:
                Column::make('Nombre','first_name')
                ->sortable(),
With
                Column::make(__('Nombre'),'first_name')
                ->sortable(),

Would allow you to easily utilise localisation files in the future.  See here for more info: https://laravel.com/docs/9.x/localization

**Reply by julianR92** [reply]
thank you for your help :)

**Reply by lrljoe** [reply]
You're more than welcome!  If you get stuck on something else Laravel or want a pointer or two, then feel free to drop me an @ in your repo and I'll see if I can point you in the right direction.

**Reply by julianR92** [reply]
thank you friend

**Comment by howdu** [comment]
> [Excluding from Column Select](https://rappasoft.com/docs/laravel-livewire-tables/v2/columns/available-methods#content-excluding-from-column-select)
> If you don't want a column to be able to be turned off from the column select box, you may exclude it:
> ```php
> Column::make('Name', 'name')
>     ->excludeFromColumnSelect()
> ```

You should be able to use this to exclude from the database column selection. That said I discovered an issue with this if you can try with this small change: https://github.com/rappasoft/laravel-livewire-tables/pull/911
```php
public function getSelectableColumns(): Collection
{
    return $this->getColumns()
        ->reject(fn (Column $column) => $column->isLabel())
        ->filter(fn (Column $column) => $column->isSelectable())
        ->values();
}
```
